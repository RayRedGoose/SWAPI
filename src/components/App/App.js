import React, { Component } from 'react'
import { Route, Switch } from 'react-router-dom'
import './App.scss'
import fakeMovies from '../../mockData/fakeMovies.js'
import { getMoviesData } from '../../apiCalls/apiCalls'
import LoginPage from '../login/LoginPage/LoginPage'
import MoviesPage from '../movies/MoviesPage/MoviesPage'
import CharactersPage from '../CharactersPage/CharactersPage'
import FavoritesPage from '../favorite/FavoritesPage/FavoritesPage'
import Page404 from '../Page404/Page404'


export default class App extends Component {
  constructor() {
    super()
    this.state = {
      user: {
        name: 'The Rock',
        quote: "I hate sand",
        rank: 'expert'},
      movies: [ ],
      favorites: [],
      sys: {
        isLoaded: false,
        error: ''
      }
    }
  }

  addMovies = () => {
    getMoviesData()
      .then(movies => {
        this.setState({
          ...this.state,
          movies,
          sys: {...this.state.sys, isLoaded: true}
        })
      })
      .catch(err => {
        this.setState({
          ...this.state,
          movies: fakeMovies,
          sys: {...this.state.sys, error: err}
        })
      })
  }

  addUser = (user) => {
     this.setState({ user })
  }

  logOut = () => {
    this.setState({ user: {} })
  }

  addFavorite = (card) => {
    this.setState({...this.state, favorites: [...this.state.favorites, card]})
  }

  removeFavorite = (name) => {
    const unFavorite =  this.state.favorites.filter(character => character.name !== name )
    this.setState({...this.state, favorites: unFavorite})
  }

  render() {
    return (
      <main className="app">
        <Switch>
          <Route exact path='/' render={() => <LoginPage addUser={this.addUser}/>} />

          <Route exact path='/movies' render={() => <MoviesPage
            sys={this.state.sys}
            user={this.state.user}
            logOut={this.logOut}
            movies={this.state.movies}
            addMovies={this.addMovies}/>} />

          <Route path='/movies/:id' render={({ match }) => {
            const specificMovie = this.state.movies.find(info => info.episode_id === parseInt(match.params.id))
            return (
              <CharactersPage
                id={match.params.id}
                movie={specificMovie}
                user={this.state.user}
                logOut={this.logOut}
                addMovies={this.addMovies}
                movies={this.state.movies}
                favorites={this.state.favorites}
                addFavorite={this.addFavorite}
                removeFavorite={this.removeFavorite}
                />
            )
          }} />

          <Route exact path='/favorites' render={() => <FavoritesPage
            user={this.state.user}
            favorites={this.state.favorites}
            removeFavorite={this.removeFavorite}
            />} />

          <Route path="*" component={Page404} />
        </Switch>
      </main>
    )
  }
}
