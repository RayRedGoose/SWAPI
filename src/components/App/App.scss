.app {
  background-color: #F1AA61;
  background-image: url('../../images/bg-image.jpg');
  background-size: cover;
  background-repeat: no-repeat;
  background-position: center;
  min-width: 100vw; min-height: 100vh;
}
