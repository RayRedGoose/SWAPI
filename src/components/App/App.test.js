import React from 'react';
import ReactDOM from 'react-dom';
import App from './App';
import characterResults from '../../mockData/fakeCharacters.js'
import movieResults from '../../mockData/fakeMovies.js'
import { shallow } from 'enzyme'
import { BrowserRouter } from 'react-router-dom'

describe("App", () => {
  let wrapper;
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<BrowserRouter>< App / ></BrowserRouter>, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  beforeEach(() => {
    wrapper = shallow(< App />, { wrapper: BrowserRouter });
  })

  it('should render App', () => {
    expect(wrapper).toMatchSnapshot();
  })

  // it('should call function to fetch api data', () => {
  //   const mockGetMoviesData = jest.fn()
  //
  //   wrapper.instance().addMovies()
  //
  //   expect(mockGetMoviesData).toHaveBeenCalled()
  // })

  it('should update state with new user when addUser is called', () => {
    const mockUser = {
      name: 'Taylor Swift',
      quote: "I find your lack of faith disturbing",
      rank: 'intermediate'};

    const expected = mockUser;

    expect(wrapper.state('user')).toEqual({
      name: 'The Rock',
      quote: "I hate sand",
      rank: 'expert'});

    wrapper.instance().addUser(mockUser);

    expect(wrapper.state('user')).toEqual(expected);
  })

  it('should update state with no user when logOut is called', () => {
    const expected = {}

    expect(wrapper.state('user')).toEqual({
      name: 'The Rock',
      quote: "I hate sand",
      rank: 'expert'});

    wrapper.instance().logOut();

    expect(wrapper.state('user')).toEqual(expected)
  })

  it('should update state with favorites array when addFavorite is called', () => {
    const mockFavoriteCharacter = characterResults[0]
    const expected = [mockFavoriteCharacter];

    expect(wrapper.state('favorites')).toEqual([])

    wrapper.instance().addFavorite(mockFavoriteCharacter);

    expect(wrapper.state('favorites')).toEqual(expected)
  })

  it('should remove correct character from state when removeFavorite is called', () => {
    const mockFavorites = characterResults;
    const mockUnfavorite1 = characterResults[1];
    const mockUnfavorite2 = characterResults[2];
    const expected = [
      {name: "Luke", species: "human", planet: "Tatioone", population: 20000, films: ['bambi', 'little women', 'sound of music'], isFavorite: false},
      {name: "Darth Vadar", species: "robot", planet: "Tatioone", population: 50000, films: ['punch drunk love'], isFavorite: true},
      {name: "Han", species: "human", planet: "Millenium Falcon", population: 60000, films:['cinderella'], isFavorite: false}
    ];

    wrapper.instance().setState({movies: movieResults, favorites: mockFavorites});

    wrapper.instance().removeFavorite(mockUnfavorite1.name);
    wrapper.instance().removeFavorite(mockUnfavorite2.name);

    expect(wrapper.state('favorites')).toEqual(expected)
  })

})
