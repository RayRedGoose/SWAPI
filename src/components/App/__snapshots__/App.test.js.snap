// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render App 1`] = `
<main
  className="app"
>
  <Switch>
    <Route
      exact={true}
      path="/"
      render={[Function]}
    />
    <Route
      exact={true}
      path="/movies"
      render={[Function]}
    />
    <Route
      path="/movies/:id"
      render={[Function]}
    />
    <Route
      exact={true}
      path="/favorites"
      render={[Function]}
    />
    <Route
      component={[Function]}
      path="*"
    />
  </Switch>
</main>
`;
